internal/observability/event/telemetry_filter.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package event
+
+import (
+	"errors"
+	"reflect"
+	"strings"
+
+	"google.golang.org/protobuf/proto"
+)
+
+// Any fields in these packages should be treated as atomic structures
+var coreProtoPackages = map[string]bool{
+	"anypb":           true,
+	"apipb":           true,
+	"durationpb":      true,
+	"emptypb":         true,
+	"fieldmaskpb":     true,
+	"sourcecontextpb": true,
+	"structpb":        true,
+	"timestamppb":     true,
+	"typepb":          true,
+	"wrapperspb":      true,
+}
+
+// protoFilter is a signature for a struct field validation test
+type protoFilter func(field reflect.StructField) bool
+
+// telemetryFilter checks a struct field should be included in observation telemetry data
+func telemetryFilter(field reflect.StructField) bool {
+	if field.Tag.Get("eventstream") == "observation" {
+		// log.Printf("%s is allowed by Tags (%s:\"%s\")", name, tag, tagValue)
+		return true
+	}
+	return false
+}
+
+// filterValue will preserve or zero a value based on if it is classed as observable
+func filterValue(fv reflect.Value, isObservable bool) {
+	if isObservable {
+		return // let data persist to telemetry
+	}
+
+	// check for nil value (prevent panics)
+	if fv == reflect.ValueOf(nil) {
+		return
+	}
+
+	if fv.Kind() == reflect.Ptr {
+		fv = fv.Elem()
+	}
+
+	// check to see if it's an exported struct field
+	if !fv.CanSet() {
+		return
+	}
+
+	fv.SetZero()
+
+	return
+}
+
+func packageNameFromType(field reflect.StructField) string {
+	typeSegments := strings.Split(field.Type.String(), ".")
+	pkg := strings.TrimLeft(typeSegments[0], "*")
+	return pkg
+}
+
+func recurseStructureWithProtoFilter(value reflect.Value, filterFunc protoFilter, isObservable bool) error {
+	kind := value.Kind()
+
+	switch kind {
+	case reflect.Interface, reflect.Ptr:
+		value = value.Elem()
+		return recurseStructureWithProtoFilter(value, filterFunc, isObservable)
+	case reflect.Map:
+		for _, k := range value.MapKeys() {
+			mVal := value.MapIndex(k)
```

Comment:
Should we have unit tests for this code path?

Reply:
I added map tests to telemetry_filter_test.go, and also changed to MapIter() in the latest version.